Sources/MongoSwift/SDAM.swift

```diff
@@ -24,11 +33,45 @@ public struct ServerAddress: Equatable {
 
     /// Initializes a ServerAddress, using the default localhost:27017 if a host/port is not provided.
     internal init(_ hostAndPort: String = "localhost:27017") throws {
-        let parts = hostAndPort.split(separator: ":")
-        self.host = String(parts[0])
-        guard let port = UInt16(parts[1]) else {
-            throw MongoError.InternalError(message: "couldn't parse address from \(hostAndPort)")
+        guard !hostAndPort.contains("?") else {
+            throw MongoError.InvalidArgumentError(message: "\(hostAndPort) contains invalid characters")
         }
+
+        var host: String
+        var port: UInt16
+
+        // Check if host is an IPv6 literal.
+        let ipLiteralRegex = try NSRegularExpression(pattern: #"^\[(.*)\](?::([0-9]+))?$"#)
+        if hostAndPort.first == "[" {
+            guard let match = ipLiteralRegex.firstMatch(
```

Comment:
I think this might be a little easier to read if this `let` was moved to the line below to line up with the next one.
e.g.
```swift
guard
    let match = ...,
    let hostRange = ...
else {
    // ...
}
```